WooCommerce/Classes/Tools/WeightFormatter.swift

```diff
@@ -1,30 +1,69 @@
 import Foundation
 
-/// Manual weight formatting
+/// Wrapper around MeasurementFormatter with a custom fallback in case there's an unrecognized unit
 ///
 final class WeightFormatter {
 
     private let weightUnit: String
-    private var withSpace: Bool
+    private let formatter: MeasurementFormatter
 
-    init(weightUnit: String, withSpace: Bool = false) {
+    /// Initializes the formatter for a given unit
+    ///
+    /// The locale parameter is meant for testing purposes.
+    /// By default, the formatter will use the system locale.
+    init(weightUnit: String, locale: Locale? = nil) {
         self.weightUnit = weightUnit
-        self.withSpace = withSpace
+        formatter = MeasurementFormatter()
+        formatter.unitOptions = .providedUnit
+        if let locale = locale {
+            formatter.locale = locale
+        }
     }
 
-    // Returns the weight plus the unit. The weight can be zero in case the value is `nil` or empty.
+    /// Returns the formatted weight with the unit.
+    /// The weight will be assumed to be zero if `nil` or empty..
     func formatWeight(weight: String?) -> String {
-        let weight: String = unwrapWeight(weight: weight)
-        guard withSpace else {
-            return weight + weightUnit
+        let weight = coalesceWeight(weight)
+
+        guard let unit = unit(symbol: weightUnit),
+              let weightValue = Double(weight) else {
+            return fallbackFormatWeight(weight: weight)
         }
-        return weight + " " + weightUnit
+        return formatWithFormatter(weight: weightValue, unit: unit)
     }
 }
 
-// MARK: - Utils
 private extension WeightFormatter {
-    func unwrapWeight(weight: String?) -> String {
-        return (weight == nil || weight?.isEmpty == true) ? "0" : (weight ?? "0")
+    func unit(symbol weightUnit: String) -> UnitMass? {
```

Comment:
Side note: This `weightUnit` `String` should probably have been converted into an `enum` somewhere. Then again, maybe a plugin can make its own unit. 🙈